libraries/chain/controller.cpp

```diff
@@ -2300,21 +2303,53 @@ struct controller_impl {
             }
          }
 
-         snapshot->read_section<value_t>([this,&row_counter]( auto& section ) {
+         snapshot->read_section<value_t>([this,&rows_loaded]( auto& section ) {
             bool more = !section.empty();
             while(more) {
                decltype(utils)::create(db, [this, &section, &more]( auto &row ) {
                   more = section.read_row(row, db);
                });
-               row_counter.progress();
+               rows_loaded.fetch_add(1u, std::memory_order_relaxed);
             }
          });
       });
 
-      read_contract_tables_from_snapshot(snapshot, row_counter);
+      snapshot_load_ctx.post([this,&snapshot,&rows_loaded]() {
+         read_contract_tables_from_snapshot(snapshot, rows_loaded);
+      });
+
+      authorization.read_from_snapshot(snapshot, rows_loaded, snapshot_load_ctx);
+      resource_limits.read_from_snapshot(snapshot, rows_loaded, snapshot_load_ctx);
+
+      constexpr unsigned max_snapshot_load_threads = 4;
+      const unsigned snapshot_load_threads = snapshot->supports_threading() ? max_snapshot_load_threads : 1;
+
+      std::vector<std::promise<void>> thread_promises(snapshot_load_threads);
+      std::list<std::thread> threads;
+      //this scoped_exit can go away with jthread; but still marked experimental in libc++18
```

Comment:
I guess I goofed this up before making #400. I thought I tested locally with libc++18 I could use `jthread` but clearly that isn't the case.